File: DoGoodBackend/src/main/java/com/dogood/dogoodbackend/api/PostAPI.java
package com.dogood.dogoodbackend.api;

import com.dogood.dogoodbackend.api.postrequests.*;
import com.dogood.dogoodbackend.domain.posts.PostDTO;
import com.dogood.dogoodbackend.domain.posts.VolunteerPostDTO;
import com.dogood.dogoodbackend.domain.posts.VolunteeringPostDTO;
import com.dogood.dogoodbackend.domain.requests.Request;
import com.dogood.dogoodbackend.domain.volunteerings.PastExperience;
import com.dogood.dogoodbackend.service.PostService;
import com.dogood.dogoodbackend.service.Response;
import jakarta.servlet.http.HttpServletRequest;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Set;

import static com.dogood.dogoodbackend.utils.GetToken.getToken;

@RestController
@CrossOrigin
@RequestMapping("/api/posts")
public class PostAPI {

    @Autowired
    private PostService postService;

    @PostMapping("/createVolunteeringPost")
    public Response<Integer> createVolunteeringPost(@RequestBody CreateVolunteeringPostRequest createVolunteeringPostRequest, HttpServletRequest request) {
        String token = getToken(request);

        String title = createVolunteeringPostRequest.getTitle();
        String description = createVolunteeringPostRequest.getDescription();
        String actor = createVolunteeringPostRequest.getActor();
        int volunteeringId = createVolunteeringPostRequest.getVolunteeringId();
        return postService.createVolunteeringPost(token, title, description, actor, volunteeringId);
    }

    @DeleteMapping("/removeVolunteeringPost")
    public Response<Boolean> removeVolunteeringPost(@RequestParam int postId, @RequestParam String actor, HttpServletRequest request) {
        String token = getToken(request);

        return postService.removeVolunteeringPost(token, postId, actor);
    }

    @PutMapping("/editVolunteeringPost")
    public Response<Boolean> editVolunteeringPost(@RequestParam int postId, @RequestBody CreateVolunteeringPostRequest createVolunteeringPostRequest, HttpServletRequest request) {
        String token = getToken(request);

        String title = createVolunteeringPostRequest.getTitle();
        String description = createVolunteeringPostRequest.getDescription();
        String actor = createVolunteeringPostRequest.getActor();
        return postService.editVolunteeringPost(token, postId, title, description, actor);
    }

    @GetMapping("/getVolunteeringPost")
    public Response<VolunteeringPostDTO> getVolunteeringPost(@RequestParam int postId, @RequestParam String actor, HttpServletRequest request) {
        String token = getToken(request);

        return postService.getVolunteeringPost(token, postId, actor);
    }

    @GetMapping("/getVolunteerPost")
    public Response<VolunteerPostDTO> getVolunteerPost(@RequestParam int postId, @RequestParam String actor, HttpServletRequest request) {
        String token = getToken(request);

        return postService.getVolunteerPost(token, postId, actor);
    }

    @GetMapping("/getAllVolunteeringPosts")
    public Response<List<VolunteeringPostDTO>> getAllVolunteeringPosts(@RequestParam String actor, HttpServletRequest request) {
        String token = getToken(request);

        return postService.getAllVolunteeringPosts(token, actor);
    }

    @GetMapping("/getOrganizationVolunteeringPosts")
    public Response<List<VolunteeringPostDTO>> getOrganizationVolunteeringPosts(@RequestParam int orgId, @RequestParam String actor, HttpServletRequest request) {
        String token = getToken(request);

        return postService.getOrganizationVolunteeringPosts(token, orgId, actor);
    }

    @PostMapping("/joinVolunteeringRequest")
    public Response<Boolean> joinVolunteeringRequest(@RequestParam String freeText, @RequestBody GeneralRequest joinRequest, HttpServletRequest request) {
        String token = getToken(request);

        int postId = joinRequest.getId();
        String actor = joinRequest.getActor();
        return postService.joinVolunteeringRequest(token, postId, actor, freeText);
    }

    @PostMapping("/searchByKeywords")
    public Response<List<? extends PostDTO>> searchByKeywords(@RequestParam String search, @RequestBody SearchPostRequest searchPostRequest, HttpServletRequest request) {
        String token = getToken(request);

        return postService.searchByKeywords(token, search, searchPostRequest.getActor(), searchPostRequest.getAllPosts(), searchPostRequest.isVolunteering());
    }

    @PostMapping("/sortByRelevance")
    public Response<List<VolunteeringPostDTO>> sortByRelevance(@RequestBody VolunteeringPostSortRequest sortRequest, HttpServletRequest request) {
        String token = getToken(request);

        return postService.sortByRelevance(token, sortRequest.getActor(), sortRequest.getAllPosts());
    }

    @PostMapping("/sortByPopularity")
    public Response<List<VolunteeringPostDTO>> sortByPopularity(@RequestBody VolunteeringPostSortRequest sortRequest, HttpServletRequest request) {
        String token = getToken(request);

        return postService.sortByPopularity(token, sortRequest.getActor(), sortRequest.getAllPosts());
    }

    @PostMapping("/sortByPostingTime")
    public Response<List<PostDTO>> sortByPostingTime(@RequestBody PostSortRequest sortRequest, HttpServletRequest request) {
        String token = getToken(request);

        return postService.sortByPostingTime(token, sortRequest.getActor(), sortRequest.getAllPosts());
    }

    @PostMapping("/sortByLastEditTime")
    public Response<List<PostDTO>> sortByLastEditTime(@RequestBody PostSortRequest sortRequest, HttpServletRequest request) {
        String token = getToken(request);

        return postService.sortByLastEditTime(token, sortRequest.getActor(), sortRequest.getAllPosts());
    }

    @PostMapping("/filterVolunteeringPosts")
    public Response<List<VolunteeringPostDTO>> filterVolunteeringPosts(@RequestBody FilterVolunteeringPostsRequest filterPostsRequest, HttpServletRequest request) {
        String token = getToken(request);

        Set<String> categories = filterPostsRequest.getCategories();
        Set<String> skills = filterPostsRequest.getSkills();
        Set<String> cities = filterPostsRequest.getCities();
        Set<String> orgNames = filterPostsRequest.getOrganizationNames();
        Set<String> volNames = filterPostsRequest.getVolunteeringNames();
        String actor = filterPostsRequest.getActor();
        boolean isMyPosts = filterPostsRequest.getIsMyPosts();
        return postService.filterVolunteeringPosts(token, categories, skills, cities, orgNames, volNames, actor, filterPostsRequest.getAllPosts() ,isMyPosts);
    }

    @PostMapping("/filterVolunteerPosts")
    public Response<List<VolunteerPostDTO>> filterVolunteerPosts(@RequestBody FilterVolunteerPostsRequest filterPostsRequest, HttpServletRequest request) {
        String token = getToken(request);

        Set<String> categories = filterPostsRequest.getCategories();
        Set<String> skills = filterPostsRequest.getSkills();
        String actor = filterPostsRequest.getActor();
        boolean isMyPosts = filterPostsRequest.getIsMyPosts();
        return postService.filterVolunteerPosts(token, categories, skills, actor, filterPostsRequest.getAllPosts(), isMyPosts);
    }

    @GetMapping("/getAllPostsCategories")
    public Response<List<String>> getAllPostsCategories(@RequestParam String actor, HttpServletRequest request) {
        String token = getToken(request);

        return postService.getAllPostsCategories(token, actor);
    }

    @GetMapping("/getAllPostsSkills")
    public Response<List<String>> getAllPostsSkills(@RequestParam String actor, HttpServletRequest request) {
        String token = getToken(request);

        return postService.getAllPostsSkills(token, actor);
    }

    @GetMapping("/getAllVolunteerPostsCategories")
    public Response<List<String>> getAllVolunteerPostsCategories(@RequestParam String actor, HttpServletRequest request) {
        String token = getToken(request);

        return postService.getAllVolunteerPostsCategories(token, actor);
    }

    @GetMapping("/getAllVolunteerPostsSkills")
    public Response<List<String>> getAllVolunteerPostsSkills(@RequestParam String actor, HttpServletRequest request) {
        String token = getToken(request);

        return postService.getAllVolunteerPostsSkills(token, actor);
    }

    @GetMapping("/getAllPostsCities")
    public Response<List<String>> getAllPostsCities(@RequestParam String actor, HttpServletRequest request) {
        String token = getToken(request);

        return postService.getAllPostsCities(token, actor);
    }

    @GetMapping("/getAllOrganizationNames")
    public Response<List<String>> getAllOrganizationNames(@RequestParam String actor, HttpServletRequest request) {
        String token = getToken(request);

        return postService.getAllOrganizationNames(token, actor);
    }

    @GetMapping("/getAllVolunteeringNames")
    public Response<List<String>> getAllVolunteeringNames(@RequestParam String actor, HttpServletRequest request) {
        String token = getToken(request);

        return postService.getAllVolunteeringNames(token, actor);
    }

    @GetMapping("/getPostPastExperiences")
    public Response<List<PastExperience>> getPostPastExperiences(@RequestParam int postId, @RequestParam String actor, HttpServletRequest request) {
        String token = getToken(request);

        return postService.getPostPastExperiences(token, actor, postId);
    }

    @GetMapping("/getVolunteeringName")
    public Response<String> getVolunteeringName(@RequestParam int volunteeringId, @RequestParam String actor, HttpServletRequest request) {
        String token = getToken(request);

        return postService.getVolunteeringName(token, actor, volunteeringId);
    }

    @GetMapping("/getVolunteeringImages")
    public Response<List<String>> getVolunteeringImages(@RequestParam int volunteeringId, @RequestParam String actor, HttpServletRequest request) {
        String token = getToken(request);

        return postService.getVolunteeringImages(token, actor, volunteeringId);
    }

    @PostMapping("/createVolunteerPost")
    public Response<Integer> createVolunteerPost(@RequestBody CreateVolunteeringPostRequest createVolunteerPostRequest, HttpServletRequest request) {
        String token = getToken(request);

        String title = createVolunteerPostRequest.getTitle();
        String description = createVolunteerPostRequest.getDescription();
        String actor = createVolunteerPostRequest.getActor();
        return postService.createVolunteerPost(token, actor, title, description);
    }

    @DeleteMapping("/removeVolunteerPost")
    public Response<Boolean> removeVolunteerPost(@RequestParam int postId, @RequestParam String actor, HttpServletRequest request) {
        String token = getToken(request);

        return postService.removeVolunteerPost(token, actor, postId);
    }

    @PutMapping("/editVolunteerPost")
    public Response<Boolean> editVolunteerPost(@RequestParam int postId, @RequestBody CreateVolunteeringPostRequest createVolunteeringPostRequest, HttpServletRequest request) {
        String token = getToken(request);

        String title = createVolunteeringPostRequest.getTitle();
        String description = createVolunteeringPostRequest.getDescription();
        String actor = createVolunteeringPostRequest.getActor();
        return postService.editVolunteerPost(token, actor, postId, title, description);
    }

    @PostMapping("/sendAddRelatedUserRequest")
    public Response<Boolean> sendAddRelatedUserRequest(@RequestBody GeneralRequest assignManagerRequest, @RequestParam String username, HttpServletRequest request) {
        String token = getToken(request);

        int postId = assignManagerRequest.getId();
        String actor = assignManagerRequest.getActor();
        return postService.sendAddRelatedUserRequest(token, actor, postId, username);
    }

    @PostMapping("/handleAddRelatedUserRequest")
    public Response<Boolean> handleAddRelatedUserRequest(@RequestBody GeneralRequest handleManagerRequest, @RequestParam boolean approved, HttpServletRequest request) {
        String token = getToken(request);

        int postId = handleManagerRequest.getId();
        String actor = handleManagerRequest.getActor();
        return postService.handleAddRelatedUserRequest(token, actor, postId, approved);
    }

    @DeleteMapping("/removeRelatedUser")
    public Response<Boolean> removeRelatedUser(@RequestParam int postId, @RequestParam String actor, @RequestParam String username, HttpServletRequest request) {
        String token = getToken(request);

        return postService.removeRelatedUser(token, actor, postId, username);
    }

    @PostMapping("/addImage")
    public Response<Boolean> addImage(@RequestParam int postId, @RequestParam String actor, @RequestBody String path, HttpServletRequest request) {
        String token = getToken(request);

        return postService.addImage(token, actor, postId, path);
    }

    @DeleteMapping("/removeImage")
    public Response<Boolean> removeImage(@RequestParam int postId, @RequestParam String actor, @RequestParam String path, HttpServletRequest request) {
        String token = getToken(request);

        return postService.removeImage(token, actor, postId, path);
    }

    @GetMapping("/getAllVolunteerPosts")
    public Response<List<VolunteerPostDTO>> getAllVolunteerPosts(@RequestParam String actor, HttpServletRequest request) {
        String token = getToken(request);

        return postService.getAllVolunteerPosts(token, actor);
    }

    @GetMapping("/getVolunteerPostRequests")
    public Response<List<Request>> getVolunteerPostRequests(@RequestParam String actor, HttpServletRequest request) {
        String token = getToken(request);

        return postService.getUserRequests(token, actor);
    }
}

File: dogood-frontend/src/api/post_api.ts
import APIResponse from "../models/APIResponse";
import axios from "axios";
import { VolunteeringPostModel } from "../models/VolunteeringPostModel";
import PastExperienceModel from "../models/PastExpreienceModel";
import { host } from "./general";
import { PostModel } from "../models/PostModel";
import { VolunteerPostModel } from "../models/VolunteerPostModel";
import RequestModel from "../models/RequestModel";

const server: string = `${host}/api/posts`;

export const createVolunteeringPost = async (title: string, description: string, volunteeringId: number): Promise<number> => {
    let username: string | null = localStorage.getItem("username");
    let token: string | null = localStorage.getItem("token");

    if(username === null) {
        throw new Error("Error");
    }

    let url = `${server}/createVolunteeringPost`;

    const config = {
        headers: { Authorization: `Bearer ${token}` }
    };
    const request = {
        title: title,
        description: description,
        volunteeringId: volunteeringId,
        actor: username
    }
    console.log(request)

    let res = await axios.post(url, request, config);
    const response: APIResponse<number> = await res.data;
    if(response.error){
        throw response.errorString;
    }
    let postId = response.data;
    return postId;
}

export const createVolunteerPost = async (title: string, description: string): Promise<number> => {
    let username: string | null = localStorage.getItem("username");
    let token: string | null = localStorage.getItem("token");

    if(username === null) {
        throw new Error("Error");
    }

    let url = `${server}/createVolunteerPost`;

    const config = {
        headers: { Authorization: `Bearer ${token}` }
    };
    const request = {
        title: title,
        description: description,
        actor: username
    }
    console.log(request)

    let res = await axios.post(url, request, config);
    const response: APIResponse<number> = await res.data;
    if(response.error){
        throw response.errorString;
    }
    let postId = response.data;
    return postId;
}

export const removeVolunteeringPost = async (postId: number) => {
    let username: string | null = localStorage.getItem("username");
    let token: string | null = localStorage.getItem("token");

    if(username === null) {
        throw new Error("Error");
    }
    
    let url = `${server}/removeVolunteeringPost?postId=${postId}&actor=${username}`;

    const config = {
        headers: { Authorization: `Bearer ${token}` }
    };
    let res = await axios.delete(url, config);
    const response: APIResponse<number> = await res.data;
    if(response.error){
        throw response.errorString;
    }
}

export const removeVolunteerPost = async (postId: number) => {
    let username: string | null = localStorage.getItem("username");
    let token: string | null = localStorage.getItem("token");

    if(username === null) {
        throw new Error("Error");
    }
    
    let url = `${server}/removeVolunteerPost?postId=${postId}&actor=${username}`;

    const config = {
        headers: { Authorization: `Bearer ${token}` }
    };
    let res = await axios.delete(url, config);
    const response: APIResponse<number> = await res.data;
    if(response.error){
        throw response.errorString;
    }
}

export const editVolunteeringPost = async (postId: number, title: string, description: string) => {
    let username: string | null = localStorage.getItem("username");
    let token: string | null = localStorage.getItem("token");

    if(username === null) {
        throw new Error("Error");
    }
    
    let url = `${server}/editVolunteeringPost?postId=${postId}`;

    const config = {
        headers: { Authorization: `Bearer ${token}` }
    };
    const request = {
        title: title, 
        description: description, 
        actor: username,
        volunteeringId: -1
    }
    let res = await axios.put(url, request, config);
    const response: APIResponse<number> = await res.data;
    if(response.error){
        throw response.errorString;
    }
}

export const editVolunteerPost = async (postId: number, title: string, description: string) => {
    let username: string | null = localStorage.getItem("username");
    let token: string | null = localStorage.getItem("token");

    if(username === null) {
        throw new Error("Error");
    }
    
    let url = `${server}/editVolunteerPost?postId=${postId}`;

    const config = {
        headers: { Authorization: `Bearer ${token}` }
    };
    const request = {
        title: title, 
        description: description, 
        actor: username
    }
    let res = await axios.put(url, request, config);
    const response: APIResponse<number> = await res.data;
    if(response.error){
        throw response.errorString;
    }
}

export const getVolunteeringPost = async (postId: number): Promise<VolunteeringPostModel> => {
    let username: string | null = localStorage.getItem("username");
    let token: string | null = localStorage.getItem("token");

    if(username === null) {
        throw new Error("Error");
    }

    let url = `${server}/getVolunteeringPost?postId=${postId}&actor=${username}`;

    const config = {
        headers: { Authorization: `Bearer ${token}` }
    };

    let res = await axios.get(url, config);
    const response: APIResponse<VolunteeringPostModel> = await res.data;
    if(response.error){
        throw response.errorString;
    }
    let post: VolunteeringPostModel = response.data;
    return post;
}

export const getVolunteerPost = async (postId: number): Promise<VolunteerPostModel> => {
    let username: string | null = localStorage.getItem("username");
    let token: string | null = localStorage.getItem("token");

    if(username === null) {
        throw new Error("Error");
    }

    let url = `${server}/getVolunteerPost?postId=${postId}&actor=${username}`;

    const config = {
        headers: { Authorization: `Bearer ${token}` }
    };

    let res = await axios.get(url, config);
    const response: APIResponse<VolunteerPostModel> = await res.data;
    if(response.error){
        throw response.errorString;
    }
    let post: VolunteerPostModel = response.data;
    return post;
}

export const getAllVolunteeringPosts = async (): Promise<VolunteeringPostModel[]> => {
    let username: string | null = localStorage.getItem("username");
    let token: string | null = localStorage.getItem("token");

    if(username === null) {
        throw new Error("Error");
    }

    let url = `${server}/getAllVolunteeringPosts?actor=${username}`;
    console.log(url);
    const config = {
        headers: { Authorization: `Bearer ${token}` }
    };

    let res = await axios.get(url, config);
    const response: APIResponse<VolunteeringPostModel[]> = await res.data;
    if(response.error){
        throw response.errorString;
    }
    let posts: VolunteeringPostModel[] = response.data;
    return posts;
}

export const getOrganizationVolunteeringPosts = async (organizationId: number): Promise<VolunteeringPostModel[]> => {
    let username: string | null = localStorage.getItem("username");
    let token: string | null = localStorage.getItem("token");

    if(username === null) {
        throw new Error("Error");
    }

    let url = `${server}/getOrganizationVolunteeringPosts?orgId=${organizationId}&actor=${username}`;

    const config = {
        headers: { Authorization: `Bearer ${token}` }
    };

    let res = await axios.get(url, config);
    const response: APIResponse<VolunteeringPostModel[]> = await res.data;
    if(response.error){
        throw response.errorString;
    }
    let posts: VolunteeringPostModel[] = response.data;
    return posts;
}

export const joinVolunteeringRequest = async (postId: number, freeText: string) => {
    let username: string | null = localStorage.getItem("username");
    let token: string | null = localStorage.getItem("token");

    if(username === null) {
        throw new Error("Error");
    }

    let url = `${server}/joinVolunteeringRequest?freeText=${freeText}`;

    const config = {
        headers: { Authorization: `Bearer ${token}` }
    };
    const request = {
        id: postId,
        actor: username
    }

    let res = await axios.post(url, request, config);
    const response: APIResponse<VolunteeringPostModel[]> = await res.data;
    if(response.error){
        throw response.errorString;
    }
}

export const sendAddRelatedUserRequest = async (postId: number, newUsername: string) => {
    console.log(newUsername);
    let username: string | null = localStorage.getItem("username");
    let token: string | null = localStorage.getItem("token");

    if(username === null) {
        throw new Error("Error");
    }

    let url = `${server}/sendAddRelatedUserRequest?username=${newUsername}`;

    const config = {
        headers: { Authorization: `Bearer ${token}` }
    };
    const request = {
        id: postId,
        actor: username
    }

    let res = await axios.post(url, request, config);
    const response: APIResponse<Boolean> = await res.data;
    if(response.error){
        throw response.errorString;
    }
}

export const searchByKeywords = async (search: string, postsToSearch: PostModel[], volunteering: boolean): Promise<PostModel[]> => {
    let username: string | null = localStorage.getItem("username");
    let token: string | null = localStorage.getItem("token");

    if(username === null) {
        throw new Error("Error");
    }

    let url = `${server}/searchByKeywords?search=${search}`;
    console.log(url)

    const config = {
        headers: { 
            Authorization: `Bearer ${token}`,
            "Content-Type": "application/json"
        }
    };
    const request = {
        actor: username,
        allPosts: postsToSearch,
        volunteering: volunteering
    }
    console.log(request)
    let res = await axios.post(url, request, config);
    const response: APIResponse<PostModel[]> = await res.data;
    if(response.error){
        throw response.errorString;
    }
    let posts: PostModel[] = response.data;
    return posts;
}

export const sortByRelevance = async (postsToSort: VolunteeringPostModel[]): Promise<VolunteeringPostModel[]> => {
    let username: string | null = localStorage.getItem("username");
    let token: string | null = localStorage.getItem("token");

    if(username === null) {
        throw new Error("Error");
    }

    let url = `${server}/sortByRelevance`;

    const config = {
        headers: { Authorization: `Bearer ${token}` }
    };
    const request = {
        actor: username,
        allPosts: postsToSort
    }
    let res = await axios.post(url, request, config);
    const response: APIResponse<VolunteeringPostModel[]> = await res.data;
    if(response.error){
        throw response.errorString;
    }
    let posts: VolunteeringPostModel[] = response.data;
    return posts;
}

export const sortByPopularity = async (postsToSort: VolunteeringPostModel[]): Promise<VolunteeringPostModel[]> => {
    let username: string | null = localStorage.getItem("username");
    let token: string | null = localStorage.getItem("token");

    if(username === null) {
        throw new Error("Error");
    }

    let url = `${server}/sortByPopularity`;

    const config = {
        headers: { Authorization: `Bearer ${token}` }
    };
    const request = {
        actor: username,
        allPosts: postsToSort
    }
    let res = await axios.post(url, request, config);
    const response: APIResponse<VolunteeringPostModel[]> = await res.data;
    if(response.error){
        throw response.errorString;
    }
    let posts: VolunteeringPostModel[] = response.data;
    return posts;
}

export const sortByPostingTime = async (postsToSort: PostModel[]): Promise<PostModel[]> => {
    let username: string | null = localStorage.getItem("username");
    let token: string | null = localStorage.getItem("token");

    if(username === null) {
        throw new Error("Error");
    }

    let url = `${server}/sortByPostingTime`;

    const config = {
        headers: { Authorization: `Bearer ${token}` }
    };
    const request = {
        actor: username,
        allPosts: postsToSort
    }
    let res = await axios.post(url, request, config);
    const response: APIResponse<PostModel[]> = await res.data;
    if(response.error){
        throw response.errorString;
    }
    let posts: PostModel[] = response.data;
    return posts;
}

export const sortByLastEditTime = async (postsToSort: PostModel[]): Promise<PostModel[]> => {
    let username: string | null = localStorage.getItem("username");
    let token: string | null = localStorage.getItem("token");

    if(username === null) {
        throw new Error("Error");
    }

    let url = `${server}/sortByLastEditTime`;

    const config = {
        headers: { Authorization: `Bearer ${token}` }
    };
    const request = {
        actor: username,
        allPosts: postsToSort
    }
    let res = await axios.post(url, request, config);
    const response: APIResponse<PostModel[]> = await res.data;
    if(response.error){
        throw response.errorString;
    }
    let posts: PostModel[] = response.data;
    return posts;
}

export const filterVolunteeringPosts = async (categories: string[], skills: string[], cities: string[], organizationNames: string[], voluntteringNames: string[], postsToFilter: VolunteeringPostModel[]): Promise<VolunteeringPostModel[]> => {
    let username: string | null = localStorage.getItem("username");
    let token: string | null = localStorage.getItem("token");

    if(username === null) {
        throw new Error("Error");
    }

    let url = `${server}/filterVolunteeringPosts`;

    const config = {
        headers: { Authorization: `Bearer ${token}` }
    };
    const request = {
        categories: categories,
        skills: skills,
        cities: cities,
        organizationNames: organizationNames, 
        volunteeringNames: voluntteringNames,
        actor: username,
        allPosts: postsToFilter
    }

    let res = await axios.post(url, request, config);
    const response: APIResponse<VolunteeringPostModel[]> = await res.data;
    if(response.error){
        throw response.errorString;
    }
    let posts: VolunteeringPostModel[] = response.data;
    return posts;
}

export const filterVolunteerPosts = async (categories: string[], skills: string[], postsToFilter: VolunteerPostModel[]): Promise<VolunteerPostModel[]> => {
    let username: string | null = localStorage.getItem("username");
    let token: string | null = localStorage.getItem("token");

    if(username === null) {
        throw new Error("Error");
    }

    let url = `${server}/filterVolunteerPosts`;

    const config = {
        headers: { Authorization: `Bearer ${token}` }
    };
    const request = {
        categories: categories,
        skills: skills,
        actor: username,
        allPosts: postsToFilter
    }

    let res = await axios.post(url, request, config);
    const response: APIResponse<VolunteerPostModel[]> = await res.data;
    if(response.error){
        throw response.errorString;
    }
    let posts: VolunteerPostModel[] = response.data;
    return posts;
}

export const getAllVolunteeringPostsCategories = async (): Promise<string[]> => {
    let username: string | null = localStorage.getItem("username");
    let token: string | null = localStorage.getItem("token");

    if(username === null) {
        throw new Error("Error");
    }

    let url = `${server}/getAllPostsCategories?actor=${username}`;

    const config = {
        headers: { Authorization: `Bearer ${token}` }
    };

    let res = await axios.get(url, config);
    const response: APIResponse<string[]> = await res.data;
    if(response.error){
        throw response.errorString;
    }
    let categories: string[] = response.data;
    return categories;
}

export const getAllVolunteeringPostsSkills = async (): Promise<string[]> => {
    let username: string | null = localStorage.getItem("username");
    let token: string | null = localStorage.getItem("token");

    if(username === null) {
        throw new Error("Error");
    }

    let url = `${server}/getAllPostsSkills?actor=${username}`;

    const config = {
        headers: { Authorization: `Bearer ${token}` }
    };

    let res = await axios.get(url, config);
    const response: APIResponse<string[]> = await res.data;
    if(response.error){
        throw response.errorString;
    }
    let skills: string[] = response.data;
    return skills;
}

export const getAllVolunteerPostsCategories = async (): Promise<string[]> => {
    let username: string | null = localStorage.getItem("username");
    let token: string | null = localStorage.getItem("token");

    if(username === null) {
        throw new Error("Error");
    }

    let url = `${server}/getAllVolunteerPostsCategories?actor=${username}`;

    const config = {
        headers: { Authorization: `Bearer ${token}` }
    };

    let res = await axios.get(url, config);
    const response: APIResponse<string[]> = await res.data;
    if(response.error){
        throw response.errorString;
    }
    let categories: string[] = response.data;
    return categories;
}

export const getAllVolunteerPostsSkills = async (): Promise<string[]> => {
    let username: string | null = localStorage.getItem("username");
    let token: string | null = localStorage.getItem("token");

    if(username === null) {
        throw new Error("Error");
    }

    let url = `${server}/getAllVolunteerPostsSkills?actor=${username}`;

    const config = {
        headers: { Authorization: `Bearer ${token}` }
    };

    let res = await axios.get(url, config);
    const response: APIResponse<string[]> = await res.data;
    if(response.error){
        throw response.errorString;
    }
    let skills: string[] = response.data;
    return skills;
}

export const getAllPostsCities = async (): Promise<string[]> => {
    let username: string | null = localStorage.getItem("username");
    let token: string | null = localStorage.getItem("token");

    if(username === null) {
        throw new Error("Error");
    }

    let url = `${server}/getAllPostsCities?actor=${username}`;

    const config = {
        headers: { Authorization: `Bearer ${token}` }
    };

    let res = await axios.get(url, config);
    const response: APIResponse<string[]> = await res.data;
    if(response.error){
        throw response.errorString;
    }
    let cities: string[] = response.data;
    return cities;
}

export const getAllOrganizationNames = async (): Promise<string[]> => {
    let username: string | null = localStorage.getItem("username");
    let token: string | null = localStorage.getItem("token");

    if(username === null) {
        throw new Error("Error");
    }

    let url = `${server}/getAllOrganizationNames?actor=${username}`;

    const config = {
        headers: { Authorization: `Bearer ${token}` }
    };

    let res = await axios.get(url, config);
    const response: APIResponse<string[]> = await res.data;
    if(response.error){
        throw response.errorString;
    }
    let cities: string[] = response.data;
    return cities;
}

export const getAllVolunteeringNames = async (): Promise<string[]> => {
    let username: string | null = localStorage.getItem("username");
    let token: string | null = localStorage.getItem("token");

    if(username === null) {
        throw new Error("Error");
    }

    let url = `${server}/getAllVolunteeringNames?actor=${username}`;

    const config = {
        headers: { Authorization: `Bearer ${token}` }
    };

    let res = await axios.get(url, config);
    const response: APIResponse<string[]> = await res.data;
    if(response.error){
        throw response.errorString;
    }
    let cities: string[] = response.data;
    return cities;
}

export const getPostPastExperiences = async (postId : number): Promise<PastExperienceModel[]> => {
    let username: string | null = localStorage.getItem("username");
    let token: string | null = localStorage.getItem("token");

    if(username === null) {
        throw new Error("Error");
    }

    let url = `${server}/getPostPastExperiences?postId=${postId}&actor=${username}`;

    const config = {
        headers: { Authorization: `Bearer ${token}` }
    };

    let res = await axios.get(url, config);
    const response: APIResponse<PastExperienceModel[]> = await res.data;
    if(response.error){
        throw response.errorString;
    }
    let experiences: PastExperienceModel[] = response.data;
    return experiences;
}

export const getVolunteeringName = async (volunteeringId : number): Promise<string> => {
    let username: string | null = localStorage.getItem("username");
    let token: string | null = localStorage.getItem("token");

    if(username === null) {
        throw new Error("Error");
    }

    let url = `${server}/getVolunteeringName?volunteeringId=${volunteeringId}&actor=${username}`;

    const config = {
        headers: { Authorization: `Bearer ${token}` }
    };

    let res = await axios.get(url, config);
    const response: APIResponse<string> = await res.data;
    if(response.error){
        throw response.errorString;
    }
    let name: string = response.data;
    return name;
}

export const getVolunteeringImages = async (volunteeringId : number): Promise<string[]> => {
    let username: string | null = localStorage.getItem("username");
    let token: string | null = localStorage.getItem("token");

    if(username === null) {
        throw new Error("Error");
    }

    let url = `${server}/getVolunteeringImages?volunteeringId=${volunteeringId}&actor=${username}`;

    const config = {
        headers: { Authorization: `Bearer ${token}` }
    };

    let res = await axios.get(url, config);
    const response: APIResponse<string[]> = await res.data;
    if(response.error){
        throw response.errorString;
    }
    let images: string[] = response.data;
    return images;
}

export const getAllVolunteerPosts = async (): Promise<VolunteerPostModel[]> => {
    let username: string | null = localStorage.getItem("username");
    let token: string | null = localStorage.getItem("token");

    if(username === null) {
        throw new Error("Error");
    }

    let url = `${server}/getAllVolunteerPosts?actor=${username}`;
    console.log(url);
    const config = {
        headers: { Authorization: `Bearer ${token}` }
    };

    let res = await axios.get(url, config);
    const response: APIResponse<VolunteerPostModel[]> = await res.data;
    if(response.error){
        throw response.errorString;
    }
    let posts: VolunteerPostModel[] = response.data;
    return posts;
}

export const getVolunteerPostRequests = async (): Promise<RequestModel[]> => {
    let username: string | null = localStorage.getItem("username");
    let token: string | null = localStorage.getItem("token");
    
    let url = `${server}/getVolunteerPostRequests?actor=${username}`;

    const config = {
        headers: { Authorization: `Bearer ${token}` }
    };
    let res = await axios.get(url, config);
    const response: APIResponse<RequestModel[]> = await res.data;
    if(response.error){
        throw response.errorString;
    }
    return response.data;
}

export const handleJoinVolunteerPostRequest = async (postId: number, approved: boolean) => {
    let username: string | null = localStorage.getItem("username");
    let token: string | null = localStorage.getItem("token");

    if(username === null) {
        throw new Error("Error");
    }
    
    let url = `${server}/handleAddRelatedUserRequest?approved=${approved}`;

    const config = {
        headers: { Authorization: `Bearer ${token}` }
    };
    const request = {
        id: postId,
        actor: username
    }

    let res = await axios.post(url, request, config);
    const response: APIResponse<number> = await res.data;
    
    if(response.error){
        throw response.errorString;
    }
}

export const addImageToVolunteerPost = async (postId: number, image: string) => {
    let username: string | null = localStorage.getItem("username");
    let token: string | null = localStorage.getItem("token");

    if (username === null) {
        throw new Error("Error");
    }

    let url = `${server}/addImage?postId=${postId}&actor=${username}`;

    const config = {
        headers: { 'Content-Type': 'application/json', Authorization: `Bearer ${localStorage.getItem('token')}` }
    };
    console.log(config);

    let res = await axios.post(url, image.replace(/"/g, ""), config);
    const response: APIResponse<boolean> = await res.data;
    if (response.error) {
        throw response.errorString;
    }
}

export const removeImageFromVolunteerPost = async (postId: number, image: string) => {
    let username: string | null = localStorage.getItem("username");
    let token: string | null = localStorage.getItem("token");

    if (username === null) {
        throw new Error("Error");
    }

    let url = `${server}/removeImage?postId=${postId}&path=${image.replace(/"/g, "")}&actor=${username}`;

    const config = {
        headers: { Authorization: `Bearer ${token}` }
    };

    let res = await axios.delete(url, config);
    const response: APIResponse<boolean> = await res.data;
    if (response.error) {
        throw response.errorString;
    }
}